### PR TITLE
Protect client secret

### LINKED ISSUE #2: Make `CLIENT_SECRET` visible after the client enters their password
`CLIENT_SECRET` is a very sensitive information and must be protected at all costs. The reason being, if compromised, the attacker can make their own app and use it to steal data from the service providing `OAuth` if the client's app is approved. A nice feature would be to hide the `CLIENT_SECRET` and require the client (or all other team members) to enter their password to view it.

### PR DESCRIPTION
**Pull request by gauransh7**
Fixes #2 
Created a simple password authentication process for viewing CLIENT_SECRET.

### CODE REVIEW

File: src/components/app-field.js [hunk 1]
```diff
@@ -1,64 +1,159 @@
-import React from 'react'
-import { connect } from 'react-redux'
-import { Table, Icon, Popup } from 'semantic-ui-react'
-
-import main from '../css/app-field.css'
+import React from "react";
```
Comment (algomaster99) [inline src/components/app-field.js]: Make sure you don't change the formatting. Revert `""`.

File: src/components/app-field.js [hunk 2]
```diff
@@ -1,64 +1,159 @@
...
-  handleClick = () => {
-    let range = document.createRange()
-    range.selectNodeContents(this.textArea)
-    let sel = window.getSelection()
-    sel.removeAllRanges()
-    sel.addRange(range)
-    document.execCommand('copy')
-    sel.removeAllRanges()
+  constructor(props) {
+    super(props);
```
Comment (algomaster99) [inline src/components/app-field.js]: Don't add semicolon

File: src/components/app-field.js [hunk 3]
```diff
@@ -1,64 +1,159 @@
...
+    this.state = {
+      password_verified: this.props.password_verified,
+      formOpen : false,
+      password : "",
+      hiddendata : "",
```
Comment (algomaster99) [inline src/components/app-field.js]: `hiddenData`. Use `camelCase ` in JS

File: src/components/app-field.js [hunk 4]
```diff
@@ -1,64 +1,159 @@
...
+      message : ""
+    };
+  }
+  handleOpen = () => this.setState({ formOpen: true });
+
+  handleClose = () => this.setState({
+     formOpen: false,
+     message : "",
+     password : ""
+    });
```
Comment (algomaster99) [inline src/components/app-field.js]: Be careful of such unnecessary indentations

File: src/components/app-field.js [hunk 5]
```diff
@@ -1,64 +1,159 @@
...
+
+  handlePasswordChange = (e,{value}) => {
+    this.setState({
+      ...this.state,
```
Comment (algomaster99) [inline src/components/app-field.js]: No need to use spread operator here. You can just use `this.setState({ password: value })`. It will automatically take update the `value` of `password`.

File: src/components/app-field.js [hunk 6]
```diff
@@ -1,64 +1,159 @@
...
+      password : value
+    })
+  }
+
+  handleSubmitSecret = (e) => {
+     e.preventDefault();
+
+     const passwordjson = {
```
Comment (algomaster99) [inline src/components/app-field.js]: Just use `this.state.password` wherever you require this.

File: src/components/app-field.js [hunk 7]
```diff
@@ -1,64 +1,159 @@
...
+       password : this.state.password,
+     }
+
+     const headers = {
+      'X-CSRFToken': getCookie('csrftoken'),
+      'content-type' : 'application/json'
+    }
+
+     axios.post(urlAppHiddenDetail(this.props.activeApp.data['id']),passwordjson,{headers: headers})
```
Comment (algomaster99) [inline src/components/app-field.js]: Break lines at 80.

File: src/components/app-field.js [hunk 8]
```diff
@@ -1,64 +1,159 @@
...
+                <Form.Input
+                  type={"password"}
+                  value={this.state.password}
+                  onChange={this.handlePasswordChange}
+                  placeholder="Enter Password"
+                />
+                <Button size="small" positive type="submit">Submit</Button>
+                <Button size="small" onClick={this.handleClose}>Cancel</Button>
+                {this.state.message=="" ? '': <Label basic color='red'>
+                    {this.state.message}
```
Comment (algomaster99) [inline src/components/app-field.js]: Extra indentation

**Review by algomaster99 — CHANGES_REQUESTED** [review]
@gauransh7 haven't tested the functionalities yet as you shared the screencast with me. Anyway, more or less there are formatting changes. For that please read the code, especially [`omniport-frontend`](https://github.com/IMGIITRoorkee/omniport-frontend), and check how different pieces of code are written.

Work on the commit messages as well before I merge this PR. Overall, the changes look great :tada:

**Comment by gauransh7** [comment]
@algomaster99 Thank you for guiding me through the iterations.
I hope the new commits serve their purpose well, and the commit messages are self explanatory.

File: src/components/app-field.js [hunk 9]
```diff
@@ -1,64 +1,157 @@
-import React from 'react'
-import { connect } from 'react-redux'
-import { Table, Icon, Popup } from 'semantic-ui-react'
-
-import main from '../css/app-field.css'
+import React from 'react';
+import { connect } from 'react-redux';
```
Comment (algomaster99) [inline src/components/app-field.js]: Group them as well.
```js
import React from 'react'
import { connect } from 'react-redux'
import Container from 'semantic-ui-react'
import { isBrowser } from 'react-device-detect'

import CustomBreadcrumb from 'core/common/src/components/custom-breadcrumb'
import { getTheme } from 'formula_one'

import RedirectURIs from './redirect-uris'
import AppField from './app-field'

import main from '../css/app-detail.css'
```

1. Third party libraries (eg. react)
2. Components of other apps in omniport ecosystem (eg. formula_one)
3. Components in this app (developer)
4. CSS of this app

File: src/components/app-field.js [hunk 10]
```diff
@@ -1,64 +1,157 @@
...
+import { Table, Icon, Popup, Button, Label, Form, FormField } from 'semantic-ui-react';
+import main from '../css/app-field.css';
+import axios from 'axios';
+import { getCookie } from 'formula_one'
+import {urlAppHiddenDetail} from '../urls'
```
Comment (algomaster99) [inline src/components/app-field.js]: Add space after `{` and before `}` - `{ urlAppHiddenDetail }`

File: src/components/app-field.js [hunk 11]
```diff
@@ -1,64 +1,157 @@
-import React from 'react'
-import { connect } from 'react-redux'
-import { Table, Icon, Popup } from 'semantic-ui-react'
-
-import main from '../css/app-field.css'
+import React from 'react';
+import { connect } from 'react-redux';
+import { Table, Icon, Popup, Button, Label, Form, FormField } from 'semantic-ui-react';
```
Comment (algomaster99) [inline src/components/app-field.js]: Remove semicolon everywhere

File: src/components/app-field.js (hunk 4 above)
Comment (algomaster99) [reply src/components/app-field.js]: This hasn't been fixed yet

File: src/components/app-field.js [hunk 12]
```diff
@@ -1,64 +1,157 @@
...
-    const { activeApp, field, verboseName, editable } = this.props
-    const { data } = activeApp
+
+  handleSubmitSecret = (e) => {
+     e.preventDefault();
+
+     const headers = {
+      'X-CSRFToken': getCookie('csrftoken'),
+      'content-type' : 'application/json'
+    }
```
Comment (algomaster99) [inline src/components/app-field.js:59]: Incorrect indentation

File: src/components/app-field.js [hunk 13]
```diff
@@ -1,64 +1,157 @@
...
+    } = this.props;
+    const { data } = activeApp;
     return (
       <Table.Row>
         <Table.Cell>{verboseName}</Table.Cell>
+        {this.state.password_verified ? (
+          <Table.Cell>
+            <div styleName="desc-container">
+              <code ref={(textarea) => (this.textArea = textarea)}>
+                {this.state.hiddenData == ""
```
Comment (algomaster99) [inline src/components/app-field.js]: Strings are enclosed in `''`
But **beware** - The values of JSX tags attributes are enclosed in `""`
```js
const name = 'Aman'
const component = <Table name="data" />
```

**Review by algomaster99 — CHANGES_REQUESTED** [review]
There are mostly formatting changes. You may use the [configuration here](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBLAtgBwgJxgAgCU4BDMAgMxwnXwHIdTy6AdKDbPfYfSKKOOXwBffFRr1GZGAFpGAEwCuAD1bssuAjwAqJAEYAbOABp8AST6mAChEyLMIsdVp0AznHQlYqMDMWo5Jhg1Ng5NfE9UKCcJOgA6OIB6MFdXRJJMTBkKVDgDeTiU1xCoMAMSVPwAQUyAMVz8-DhleCh5VyIguKtFRgBhGmwBWG42fHwACy95Iz6DHwBrfABefAAKAEoVgD5R6PH8IwIcLwBzOBX8eQgwRXQEGEKpeEIzuE2xg5Ooc7j3I3IADkIPI4ANYA9XGsYBNUK44vAWlUpBtPuMjvh-pcAO5Ra7YuLnGAAZTyghgqGgH3243+cUY6AgADc4FUDAZXj84FDUTTMXk4iR5PJOec1t9zryDlcbncHnFmoIBuhPG01nRIJgAJ50KUHOkM5ms9mi7nU8bCT6MNpwHDrLbANG8aCuLT4aSoFk1TCmHJ5eSmFk4PQQdyAkj3UxweSoGD6IyOVYwuFxTDUTCuJ18V3cK4kOOJ93kT2szJOxgwXrRNZO8YAHl0hjgcUIEGx21rBwb8ebfTyBm2wCDIbDEbgwjriUbRjiffZHb50u7Tdn-YX0o3+DrMaZmJgWqM4fuyzooNcvj4caitro683G7rkFB+EYFGWwERcakO3Wyfhn+RUhLk-EgUWEO97w3YB5HzEgAG0-XyABdS1F3vScnzgCD0MSHdsKXKce1XedO3radezXUiDkdNDNzrGw7EwKjNxgHBUFOc4cHfZj0IsaAeMgqAxxPTUdQE+95igBZxJY2MjBPOZFnwGAIGdbU1EgzT8GgRSwAWd8-ziKY2lmeY9NQrTIMgAxcBPU5GDE2jNMSfDNIsyzxkvB53wAAwAEiHW0RzgI9xzU3J5AAQh89yPOgESzIWDTLJcnjYofQiVznAdO0nciWzbbC9VQ1C2AoRRSgpaAIgyYk43gbQICsdMOjWV18zgB0rTgSscGiGjpQ9L1MiQPcOsFYthqY-YSrYbMCE8TAABE4UwfMwAmRrmtsDpVhjVw1pgDafwGl8eqrbhZqgNhmk4AhQQoEhFAMAg+AEcga32Ra6o6raWuMT5FpWg71s2pqWrYDY1m9ep-Q2EBjBAWwqqgVxkFAUDqGxKxQIQNGUBIAxsRILU0cRvQTj0nriTWsAolOZBWMUEwQAmGB0AMAB1WF4BBsA4FJVHY09WMtWQcBUgRtBUdtGBmpIU5PGQR6DHcRGACtXGUAAhSmFmpscABlr2Vwm1ZATXlGJemjAARUUCB4FN1WWbWnB3BwcW9H0PIpbTKIYE51B5BhZAAA4AAZEbTUM4E5k5MHFtNuVtFkpYARwd+BtozcWKhkARo2jKXGEz1BGHlxWSGd833HQVAa5Z1wbbge3HbgRmcGZxG4z0IOQ4mZAACYe5OVBJNOZUlZQbkAFYpcUdxG3xlXzaZZmzBtWBiTANjMBgKo2jqg8O6QVfx2EIA) to format your code. But you will have to add newlines manually.

File: src/components/app-field.js (hunk 2 above)
Comment (algomaster99) [reply src/components/app-field.js]: Still left

**Review by algomaster99 — CHANGES_REQUESTED** [review]
@gauransh7  just one small change. Remove semicolon from everywhere. Also, can you send the screencast here? Use ezgif.com to convert video formats to gif

**Comment by gauransh7** [comment]
CLIENT_SECRET

@algomaster99 
1. Color of **Cancel** button has been changed to default.
2. `Label postition` or **Wrong Password** has been changed.

There are no other changes in design apart from these two.

**Comment by pradumangoyal** [comment]
I might be a bit late. You can configure your VS Code according to this https://app.nuclino.com/IMGIITRoorkee/Omniport-Development/Frontend-623e56dd-54c2-428e-93e0-a52f2c030d89 and there won't be any deviation from the Omniport frontend code conventions. Once configure your VS Code according to this and press Ctrl+S on every component you edited and push that code.

File: src/components/app-field.js (hunk 13 above)
Comment (pradumangoyal) [reply src/components/app-field.js]: For JSX, we generally follows <Table name={'data'} />. But genrally constants are not passed in the JSX elements. They are generally stae or props variables.

File: src/components/app-field.js [hunk 14]
```diff
@@ -1,64 +1,180 @@
...
+  }
+
   handleClick = () => {
     let range = document.createRange()
     range.selectNodeContents(this.textArea)
     let sel = window.getSelection()
     sel.removeAllRanges()
     sel.addRange(range)
-    document.execCommand('copy')
+    document.execCommand("copy")
```
Comment (algomaster99) [inline src/components/app-field.js]: Strings are in single quote (`''`)

File: src/components/app-field.js [hunk 15]
```diff
@@ -1,64 +1,180 @@
...
     const { activeApp, field, verboseName, editable } = this.props
     const { data } = activeApp
     return (
       <Table.Row>
         <Table.Cell>{verboseName}</Table.Cell>
+        {this.state.password_verified ? (
+          <Table.Cell>
+            <div styleName="desc-container">
+              <code ref={(textarea) => (this.textArea = textarea)}>
+                {this.state.hiddenData == ''
```
Comment (algomaster99) [inline src/components/app-field.js]: Always use `===`.
image

File: src/components/app-field.js [hunk 16]
```diff
@@ -1,64 +1,180 @@
...
+              </code>
+            </div>
+          </Table.Cell>
+        ) : (
+          <Table.Cell>
+            {this.state.formOpen ? (
+              <div className="appfield_div">
+                <Input
+                  styleName="appfield_input"
+                  type={"password"}
```
Comment (algomaster99) [inline src/components/app-field.js]: When you enclose something in `{}`, you are essentially writing JS. So what's the rule for string? They should be in `''`.

**Review by algomaster99 — CHANGES_REQUESTED** [review]

File: src/components/app-field.js [hunk 17]
```diff
@@ -1,10 +1,86 @@
...
+
+  handlePasswordChange = (e, { value }) => {
+    this.setState({
+      password: value,
+    })
+  }
+
+  handleSubmitSecret = () => {
+    if(this.state.password === ''){
+      this.handleSubmitError("Enter something")
```
Comment (algomaster99) [inline src/components/app-field.js]: Reword this to `'Password cannot be blank'` (notice the quotes too)

File: src/components/app-field.js [hunk 18]
```diff
@@ -1,10 +1,86 @@
...
+      return
+    }
+
+    const headers = {
+      "X-CSRFToken": getCookie("csrftoken"),
```
Comment (algomaster99) [inline src/components/app-field.js]: Single quotes.

File: src/components/app-field.js [hunk 19]
```diff
@@ -1,10 +1,86 @@
...
+      "content-type": "application/json",
+    }
+
+    axios
+      .post(
+        urlAppHiddenDetail(),
+        {
+          password: this.state.password,
+          id: this.props.activeApp.data["id"],
```
Comment (algomaster99) [inline src/components/app-field.js]: Single quotes

File: src/components/app-field.js [hunk 20]
```diff
@@ -1,10 +1,86 @@
...
+        this.setState({
+          password_verified: true,
+          hiddenData: res.data[this.props.field],
+        })
+      })
+      .catch((err) => this.handleSubmitError(err.response.data))
+  }
+
+  handleKeyDown = (event) => {
+    if(event.key === 'Enter'){
```
Comment (algomaster99) [inline src/components/app-field.js]: It is better to use `event.keyCode === <code for Enter>`. Reference: https://www.w3schools.com/jsref/event_key_keycode.asp

File: src/components/app-field.js [hunk 21]
```diff
@@ -1,10 +1,86 @@
...
+    })
+
+  handlePasswordChange = (e, { value }) => {
+    this.setState({
+      password: value,
+    })
+  }
+
+  handleSubmitSecret = () => {
+    if(this.state.password === ''){
```
Comment (algomaster99) [inline src/components/app-field.js]: Space between `if` and `(` and `)` and `{`

**Review by algomaster99 — CHANGES_REQUESTED** [review]
Really sorry I didn't look at the reviews before. Can you push another commit? I am sure we are done then.

By the way, didn't you run it through prettier?

**Review by algomaster99 — APPROVED** [review]

File: src/css/app-field.css [hunk 22]
```diff
@@ -1,3 +1,17 @@
...
+.desc-container {
+  word-break: break-all;
+}
+
+.app-field-div {
+  display: flex;
+  flex-direction: row;
+}
+
+.app-field-input-browser {
```
Comment (algomaster99) [inline src/css/app-field.css:11]: You can smartly use selectors to eliminate redundancies.
```css
.app-field-input-browser,
.app-field-input-mobile {
  margin-right: 0.33em;
}

.app-field-input-mobile {
  margin-bottom: 0.33em;
}
```

File: src/components/app-detail.js [hunk 23]
```diff
@@ -151,11 +151,13 @@ class AppDetail extends React.Component {
                   field='clientId'
                   verboseName='Client ID'
                   editable={false}
+                  password_verified = {true}
```
Comment (algomaster99) [inline src/components/app-detail.js]: Sorry I just overlooked this file before. But don't add ` ` (space) here.

File: src/components/app-detail.js [hunk 24]
```diff
@@ -151,11 +151,13 @@ class AppDetail extends React.Component {
...
                 />
                 <AppField
                   field='clientSecret'
                   verboseName='Client secret'
                   editable={false}
+                  password_verified = {false}
```
Comment (algomaster99) [inline src/components/app-detail.js]: Here too

**Review by algomaster99 — CHANGES_REQUESTED** [review]

File: src/components/app-detail.js [hunk 25]
```diff
@@ -265,7 +265,4 @@ const mapDispatchToProps = dispatch => {
   }
 }
 
-export default connect(
-  mapStateToProps,
-  mapDispatchToProps
-)(AppDetail)
+export default connect(mapStateToProps, mapDispatchToProps)(AppDetail)
```
Comment (gauransh7) [inline src/components/app-detail.js:268]: This commit was suggested by prettier code-formatter, moreover it is feasible as other files are following the same pattern.

File: src/components/app-field.js [hunk 26]
```diff
@@ -110,21 +110,23 @@ class AppField extends React.PureComponent {
...
+                    }
+                    type={'password'}
+                    value={this.state.password}
+                    onChange={this.handlePasswordChange}
+                    placeholder={this.state.passwordPlaceholder}
+                    error={this.state.passwordError}
+                    onKeyDown={this.handleKeyDown}
+                  />
+                </div>
+                <div styleName='app-field-div'>
```
Comment (gauransh7) [inline src/components/app-field.js]: The buttons will have no margin between them if `ButtonGroup` is used.
I have modified the code so that the buttons will have **default margin** between them and the design is device friendly.

**Review by algomaster99 — APPROVED** [review]

**Comment by algomaster99** [comment]
@gauransh7 great work :tada: we are ready for merge
